backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/orchestration/impl/AbstractElasticsearchBarrierWorkOrchestrator.java

```diff
@@ -63,13 +63,29 @@ protected final String getName() {
 		}
 	}
 
+	// TODO avoid code duplication here
```

Comment:
Indeed. You could simply move the concept of "Changeset" to this class and have four methods:

* `public <T> CompletableFuture<T> submit(ElasticsearchWork<T>)` creates a `SingleWorkChangeset`, calls `submit(Changeset<T>)` and returns the future passed to the constructor of the changeset.
* `public CompletableFuture<?> submit(List<ElasticsearchWork<?>>)` creates a `MultiWorkChangeset`, calls `submit(Changeset<T>)` and returns the future passed to the constructor of the changeset.
* `private <T> void submit(Changeset<T>)` implements the code below and delegates to `doSubmit(Changeset<T>)`
* `protected abstract <T> void doSubmit(Changeset<T>)` is abstract and implemented in subclasses

Reply:
Good idea. Thank you. I will apply it.